File: 0020_valid_parentheses/solution.py
from collections import deque
class Solution:
    def isValid(self, s: str) -> bool:
        opening = "{[("
        open_close = {"{":"}", "(":")", "[":"]"}
        stack = deque()
        count = 0
        for bracket in s:
            if bracket in opening:
                count += 1
                stack.append(bracket)
            else:
                try:
                    popped = stack.pop()
                    count -=1
                except:
                    return False
                if bracket != open_close[popped]:
                    return False
        if count == 0:
            return True
        return False

File: 0020_valid_parentheses/solution.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, bool> opening{
                {'{', true},
                {'[', true},
                {'(', true}
        };
        unordered_map<char, char> open_close{
                {'{', '}'},
                {'[', ']'},
                {'(', ')'}
        };
        stack<char> stk;
        int count = 0;
        for(char& i : s){
            if(!(opening.find(i) == opening.end())){
                stk.push(i);
                count++;
            }
            else{
                if(!stk.empty()){
                    char h = stk.top();
                    stk.pop();
                    count--;
                    //if closing bracket doesn't match, return false
                    if(!(i == open_close.at(h))){
                        cout << "Unbalanced!" << endl;
                        return false;
                    }

                }
                else{
                    cout << "Unbalanced!" << endl;
                    return false;
                }
            }
        }
        if(count == 0){
            cout << "Balanced!" << endl;
            return true;
        }
        else{
            cout << "Unbalanced" << endl;
            return false;
        }
    }
};
